### PR TITLE
Implement setLifeCycleToAllIndexes in ES and OS.

### PR DESCRIPTION
**Pull request by rodrigo-lourenco-lopes**
## Description

This test adds the implementation of setLifeCycleToAllIndexes along the unit tests for OS and ES.

## Checklist

- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)

## Related issues

closes #

### CODE REVIEW

**Comment by rodrigo-lourenco-lopes** [comment]
> In this PR I am not sure about 2 details, if we should pass additionally the ConnectConfiguration to the repositories since now we also need the Index prefix in the repositories, or if its just better just to pass the entire exporter configuration, since now we are passing 3 configurations there, all from the exporter config.
The second question is that if in the both cases where we apply policies, setIndexLifeCycle() and setLifeCycleToAllIndexes() if we should make sure that the policy exists (and create if it does not)?

after discussing the topic, we decided that `setIndexLifeCycle()` and `setLifeCycleToAllIndexes()` should not check if policy exists as this gets created before synchronously.
As for the `ConnectConfiguration` there is not point passing the entire configuration if we are only using the index prefix, so we only pass this field.

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/archiver/OpenSearchRepository.java
Comment (npepinpe) [inline zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/archiver/OpenSearchRepository.java:205]: :wrench: `String#matches` will convert this to a compiled `Pattern` instance. Since our index never changes, we can already compile it ourselves once, and then just use to match.

```suggestion
  final var matcher = Pattern.compile(indexWildCard).asMatchPredicate();

  client
        .cat()
        .indices()
        .thenAccept(
            response ->
                response.valueBody().stream()
                    .map(IndicesRecord::index)
                    .filter(matcher)
```

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/archiver/OpenSearchRepository.java
Comment (npepinpe) [inline zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/archiver/OpenSearchRepository.java:207]: :x: Joining here will pin the background thread indefinitely. This applies as well to the ES repository version, which also joins the result of fetching the indices.

Instead, we should compose the operations:

```java
  @Override
  public CompletableFuture<Void> setLifeCycleToAllIndexes() {
    if (!retention.isEnabled()) {
      return CompletableFuture.completedFuture(null);
    }

    final var indexWildCard = "^" + indexPrefix + INDEX_WILDCARD;
    return fetchIndexMatchingIndexes(indexWildCard)
      .thenComposeAsync(indices -> setIndexLifeCycle(indices.toArray(String[]::new)), executor);
  }

  private CompletableFuture<List<String>> fetchIndexMatchingIndexes(final String indexWildCard) throws IOException {
    return client
        .cat()
        .indices()
        .thenAccept(
            response ->
                response.valueBody().stream()
                    .map(IndicesRecord::index)
                    .filter(index -> index.matches(indexWildCard))
                    .toList());

    return indices;
  }
```

File: zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/tasks/archiver/OpenSearchRepositoryTest.java
Comment (npepinpe) [inline zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/tasks/archiver/OpenSearchRepositoryTest.java:69]: :wrench: You can just pass some "foo" or something here, doesn't really matter.

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/archiver/ElasticsearchRepository.java
Comment (npepinpe) [inline zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/archiver/ElasticsearchRepository.java:199]: :x: Do not join to avoid pinning the thread. See the comment in the OS reposiotry class.

**Review by npepinpe — CHANGES_REQUESTED** [review]

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/BackgroundTaskManager.java
Comment (rodrigo-lourenco-lopes) [reply zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/BackgroundTaskManager.java]: Good catch!

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/archiver/OpenSearchRepository.java
Comment (rodrigo-lourenco-lopes) [reply zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/archiver/OpenSearchRepository.java]: Good point all these are returning completable futures!

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/archiver/OpenSearchArchiverRepository.java
Comment (npepinpe) [inline zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/archiver/OpenSearchArchiverRepository.java:140]: :x: Use `thenComposeAsync` and specify the executor. This ensures that we will skip the callback if the background task manager is closed.

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/archiver/ElasticsearchArchiverRepository.java
Comment (npepinpe) [inline zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/archiver/ElasticsearchArchiverRepository.java:196]: :x: Use `thenApplyAsync` and specify the executor, that way we will not execute the callback if the background task manager is shut down.

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/archiver/ElasticsearchArchiverRepository.java
Comment (npepinpe) [inline zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/archiver/ElasticsearchArchiverRepository.java:139]: :x: Use `thenComposeAsync` and specify the executor.

File: zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/archiver/OpenSearchArchiverRepository.java
Comment (npepinpe) [inline zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/tasks/archiver/OpenSearchArchiverRepository.java:213]: :x: Specify the executor as well to ensure we don't execute the callback after shutdown.

**Review by npepinpe — APPROVED** [review]
:rocket: 

Please address all comments before merging, but no need for another review

**Pull request merged**
